socorro/processor/processor_2015.py

```diff
@@ -261,51 +269,47 @@ def process_crash(self, raw_crash, raw_dumps, processed_crash):
         processed_crash.signature = 'EMPTY: crash failed to process'
 
         crash_id = raw_crash['uuid']
-        try:
-            # quit_check calls ought to be scattered around the code to allow
-            # the processor to be responsive to requests to shut down.
-            self.quit_check()
 
-            start_time = self.config.logger.info(
-                "starting transform for crash: %s",
-                crash_id
-            )
-            processor_meta_data.started_timestamp = start_time
-
-            # apply_all_rules
-            for rule in self.rules:
-                rule.act(
-                    raw_crash,
-                    raw_dumps,
-                    processed_crash,
-                    processor_meta_data
+        # quit_check calls ought to be scattered around the code to allow
+        # the processor to be responsive to requests to shut down.
+        self.quit_check()
+
+        start_time = self.config.logger.info('starting transform for crash: %s', crash_id)
+        processor_meta_data.started_timestamp = start_time
+
+        # Apply rules; if a rule fails, capture the error and continue onward
+        for rule in self.rules:
+            try:
+                rule.act(raw_crash, raw_dumps, processed_crash, processor_meta_data)
+
+            except Exception as exc:
+                # If a rule throws an error, capture it and toss it in the
+                # processor notes
+                raven_client.capture_error(
+                    sentry_dsn=self.sentry_dsn,
+                    logger=self.config.logger,
+                    extra={'crash_id': crash_id}
                 )
-                self.quit_check()
-
-            # the crash made it through the processor rules with no exceptions
-            # raised, call it a success.
-            processed_crash.success = True
-
-        except Exception as exception:
-            self.config.logger.warning(
-                'Error while processing %s: %s',
-                crash_id,
-                str(exception),
-                exc_info=True
-            )
-            processor_meta_data.processor_notes.append(
-                'unrecoverable processor error: %s' % exception
-            )
+                # NOTE(willkg): notes are public, so we can't put exception
+                # messages in them
+                processor_meta_data.processor_notes.append(
+                    'rule %s failed: %s' % (rule.__class__.__name__, exc.__class__.__name__)
```

Comment:
This creates a line like "rule OSInfoRule failed: KeyError". That gives me something to search for to reprocess crashes that had issues but shouldn't reveal any PII in a public field.

Reply:
The thing that calls `process_crash` also has error handling, but that error handling is tied in with processing the crash and saving it. I didn't want to further complicate that. Further, if we handled errors there, processing would halt on the first error thrown. That'd be great if we had a system where crashes that couldn't get through processing were put aside in a separate bin. But we don't have that. Our options are:

1. continue processing despite errors and end up with a partially processed crash
2. throw it back in the rabbitmq queue
3. drop it altogether and never process it again

We can't do option 3 because then we have crashes we've saved to S3 but don't have indexed and don't know anything about. Option 2 could lead to the queue filled with bad crashes and the processor grinding to a halt. Option 1 is unenthusing and could lead to confusion during analysis.

I'm going for option 1. The hope is that we'll catch the errors in sentry and fix the problems and push the fixes out. Since we're adding notes to the crashes in question, we can reprocess them once the problem is fixed.

We'll see how that goes.